pyquil/_memory.py

```diff
@@ -0,0 +1,56 @@
+from dataclasses import dataclass, field
+from typing import Dict, Sequence, Union
+
+from rpcq.messages import ParameterAref
+
+
+@dataclass
+class Memory:
+    """
+    Memory encapsulates the values to be sent as parameters alongside a program at time of
+    execution, and read back afterwards.
+    """
+
+    values: Dict[ParameterAref, Union[int, float]] = field(default_factory=dict)
+
+    def copy(self) -> "Memory":
+        """
+        Return a deep copy of this Memory object.
+        """
+        return Memory(values={k.replace(): v for k, v in self.values.items()})
+
+    def write(self, parameter_values: Dict[Union[str, ParameterAref], Union[int, float]]) -> "Memory":
+        """
+        Set the given values for the given parameters.
+        """
+        for parameter, parameter_value in parameter_values.items():
+            self._write_value(parameter=parameter, value=parameter_value)
+        return self
+
+    def _write_value(
+        self,
+        *,
+        parameter: Union[ParameterAref, str],
+        value: Union[int, float, Sequence[int], Sequence[float]],
+    ) -> "Memory":
+        """
+        Mutate the program to set the given parameter value.
+
+        :param ParameterAref|str parameter: Name of the memory region, or parameter reference with offset.
```

Comment:
nit: Is it necessary to add types to param docs? It seems it could go stale easily, and we have the type annotations in the signature already.

Reply:
good point